### DISCUSSION TITLE
First time user feedback

### DISCUSSION THREAD

**Discussion by beatcracker** [dagger/dagger#3991] [General]
Hello folks!

Let me start with saying that the idea behind the Dagger (composable re-usable actions) is something that resonates with me and I really want Dagger to work for my use-cases. Unfortunately there are lot of things that are unclear and/or broken which makes Dagger hard to adopt. 

> Please note that everting below is written from the point of view of the OPS person who doesn't do much software development

I first learned about Dagger when CUE was THE way to build the pipeline graph. At that point of time it seemed like a very logical approach - one declarative config to rule them all. Build the initial graph using CUE, submit it to the Dagger engine and let it figure everything else for you. Finally I could get away from re-inventing restore/build/publish steps for each bespoke CI and make them environment-agnostic.

Nowadays there are several SDKs, which look like more _imperative_ way to drive Dagger engine w/o any ability to get hold of intermediate rendered graph. Which feels like 🐔 and 🥚 problem: do I need to be able to run GO/Python/Node code in my CI system to be able to run Dagger actions? One might expect that SDK will be able to dump the final graph which can be stored in the repo/universe and re-used directly by the Dagger engine in CI.

Unfortunately, the documentation doesn't make it entirely clear for me how  GO/Python/Node SDKs work under the hood. The [CUE documentation](https://docs.dagger.io/sdk/cue/232322/guides) is still the most complete one, which makes it weird to see GO SDK pushed forward. My understanding is that CUE is to hard to adopt, so language SDKs are here to make it easier, but it feels like loosing the most important part of the initial promise: composable re-usable actions. There a lot of ways to drive a container build/runtime engine and if Dagger positions itself like this then it loses it's major advantage. There are tools like Earthly which provide much easier ways to do so.

There also a lot of stuff that is slightly broken or doesn't make much sense due to the Dagger version migrations: there is no proper changelog which outlines the major changes for going from `0.1` to `0.2` and `0.3` in the future. Moreover, the versions are not mentioned in the documentation at all, and you learn about them only while going through the GitHub issues, which mention things like `Europa` and `Cloak`, but make it hard to connect those to the actual downloads. Especially with the (new?) Dagger CLI available in [releases](https://github.com/dagger/dagger/releases) which is also not properly covered anywhere in the docs. Plus, the VSCode extension source code is nowhere to be found. 

All in all I really want to like Dagger and make use of it, but it's way harder that its should be and it would be a hard sell to my current "devops" team in it's current state.

Please take this feedback in the good faith - I'll be happy to clarify it further or answer any questions you might have.

**Comment by jpadams** [comment]
Thanks for the feedback. Some notes below.

> Nowadays there are several SDKs, which look like more _imperative_ way to drive Dagger engine w/o any ability to get hold of intermediate rendered graph. Which feels like 🐔 and 🥚 problem: do I need to be able to run GO/Python/Node code in my CI system to be able to run Dagger actions? One might expect that SDK will be able to dump the final graph which can be stored in the repo/universe and re-used directly by the Dagger engine in CI.

Yes! To run a pipeline using a Dagger Go, Python, Node.js SDK, you need to be able to run it in your CI system. Go also allows for compilation of a static binary, of course. The nice thing is that's all you need. If you can run Go or Javascript or Python, and Docker containers, then you're good to go. 🚀 


> Unfortunately, the documentation doesn't make it entirely clear for me how GO/Python/Node SDKs work under the hood. 

Does this help? https://docs.dagger.io/#how-does-it-work
You'll hear more about the underlying tech in a release soon! 😉

> There also a lot of stuff that is slightly broken or doesn't make much sense due to the Dagger version migrations: there is no proper changelog which outlines the major changes for going from `0.1` to `0.2` and `0.3` in the future.

Hope this helps: https://github.com/dagger/dagger/releases/
Guides should be based on latest. An issue or a PR is super appreciated if something is out of date. 🙏
